driver-core/src/main/java/com/datastax/driver/core/RegularStatement.java

```diff
@@ -43,31 +111,926 @@ protected RegularStatement() {}
+        return setMap(name, v, TypeToken.of(keysClass), TypeToken.of(valuesClass));
+    }
+
+    @Override
+    public <K, V> RegularStatement setMap(int i, Map<K, V> v, TypeToken<K> keysType, TypeToken<V> valuesType) {
+        return set(i, v, mapOf(keysType, valuesType));
+    }
+
+    @Override
+    public <K, V> RegularStatement setMap(String name, Map<K, V> v, TypeToken<K> keysType, TypeToken<V> valuesType) {
+        return set(name, v, mapOf(keysType, valuesType));
+    }
+
+    
+    @Override
+    public Object getObject(int i) {
+        return getInternal(i).as(null);
+    }
+
+    @Override
+    public Object getObject(String name) {
+        return getInternal(name).as(null);
+    }
+
+    
+    @Override
+    public <V> V get(int i, Class<V> targetClass) {
+        return get(i, TypeToken.of(targetClass));
+    }
+
+    @Override
+    public <V> V get(String name, Class<V> targetClass) {
+        return get(name, TypeToken.of(targetClass));
+    }
+
+    @Override
+    public <V> V get(int i, TypeToken<V> targetType) {
+        Value<V> value = getInternal(i);
+        return value.as(targetType);
+    }
+
+    @Override
+    public <V> V get(String name, TypeToken<V> targetType) {
+        Value<V> value = getInternal(name);
+        return value.as(targetType);
+    }
+
+    @Override
+    public <V> V get(int i, TypeCodec<V> codec) {
+        Value<V> value = getInternal(i);
+        return value.with(codec);
+    }
+
+    @Override
+    public <V> V get(String name, TypeCodec<V> codec) {
+        Value<V> value = getInternal(name);
+        return value.with(codec);
+    }
+
+    @Override
+    public <V> RegularStatement set(int i, V v, Class<V> targetClass) {
+        return set(i, v, TypeToken.of(targetClass));
+    }
+
+    @Override
+    public <V> RegularStatement set(String name, V v, Class<V> targetClass) {
+        return set(name, v, TypeToken.of(targetClass));
+    }
 
     @Override
-    public String toString() {
-        return getQueryString();
+    public <V> RegularStatement set(int i, V v, TypeToken<V> targetType) {
+        // we must treat tuples and UDTs separately
+        // as they convey their own specific CQL types as well
+        DataType cqlType = null;
+        if (v instanceof TupleValue)
+            cqlType = ((TupleValue)v).getType();
+        if (v instanceof UDTValue)
+            cqlType = ((UDTValue)v).getType();
+        return set(i, v, codecRegistry.codecFor(cqlType, targetType));
     }
+
+    @Override
+    public <V> RegularStatement set(String name, V v, TypeToken<V> targetType) {
+        // we must treat tuples and UDTs separately
+        // as they convey their own specific CQL types as well
+        DataType cqlType = null;
+        if (v instanceof TupleValue)
+            cqlType = ((TupleValue)v).getType();
+        if (v instanceof UDTValue)
+            cqlType = ((UDTValue)v).getType();
+        return set(name, v, codecRegistry.codecFor(cqlType, targetType));
```

Comment:
It's too bad that we forget `targetType` after that point. We could propagate it all the way to store it in the `Value` instance. That would even make the type check in `as()` (that I suggested in my other comment) much better. A raw value would be one that doesn't have a target type.